### DISCUSSION TITLE
Get sprite sizes

### DISCUSSION THREAD

**Discussion by HotSammiches** [bevyengine/bevy#2603] [Q&A]
What is the best way to get the size of a sprite once it's been loaded? For example, once I have a sprite_bundle with a given transform how can I get the sprite's x and y sizes? Sorry if this is a silly question - I haven't been able to figure it out and I really don't want to have to manually enter the size of every sprite (which would be necessary for things like collision detection).

**Comment by hymm** [comment]
you can get it from the Sprite Component https://docs.rs/bevy/0.5.0/bevy/prelude/struct.Sprite.html

**Reply by HotSammiches** [reply]
I tried that, but when I did both the x and y came back as 0. Does it have to be after spawn_bundle is called?

**Reply by hymm** [reply]
yeah.  Some system fills in the values after the bundle is spawned.

**Comment by meepleek** [comment]
I believe you should be able to get the size from the texture (well, the texture handle) after it's been loaded.

**Comment by The-DevBlog** [comment]
Can you provide some code for what you are trying to accomplish?

**Reply by HotSammiches** [reply]
For example, this gets me what I want:
```
    let texture_handle = asset_server.load(PLAYER_SPRITE);
    let texture_atlas = TextureAtlas::from_grid(
        texture_handle,
        Vec2::new(PLAYER_SPRITE_TILE_SIZE_X, PLAYER_SPRITE_TILE_SIZE_Y),
        PLAYER_SPRITE_TILE_COLS,
        PLAYER_SPRITE_TILE_ROWS,
    );
    let sprite_size = texture_atlas.size;

    println!("Player sprite size: {} {}", sprite_size.y, sprite_size.x);
```

I can then access sprite_size.x and sprite_size.y and they are equal to PLAYER_SPRITE_TILE_SIZE_X and PLAYER_SPRITE_TILE_SIZE_Y (both of which are constants). I know this is necessary for a sprite sheet since it contains many images in one file, but if I want to just load a single sprite I still have to manually enter the size of the sprite in order to have the x and y sizes. Here's an example:

```
    let texture_handle = asset_server.load("sprites/table.png");
    let sprite_bundle = SpriteBundle {
        material: materials.add(texture_handle.into()),
        ..Default::default()
    };
    let sprite_size = sprite_bundle.sprite.size;

    commands.spawn_bundle(sprite_bundle);

    println!("Table CB: {} {}", sprite_size.y, sprite_size.x);
```

The output of a program which runs both of these code snippets is:

```
Player sprite size: 44 72
Table CB: 0 0
```

What would be really nice is to be able to automatically populate the sprite size in the same function. Failing that it would be acceptable to be able to run a query after the fact to pull the sizes in a subsequent system.

The concern I have is that manually retyping a sprite's size if I need to resize it is error-prone. I want to automate that so I don't have to remember as much.

**Reply by The-DevBlog** [reply]
Are you saying you want to figure out how to dynamically calculate the size of the individual sprite frames from the sprite sheet and store that information somewhere?

**Reply by HotSammiches** [reply]
Kind of - I don't want it automatically from the sprite sheet but from Bevy after the sprite is spawned. And not just sprites from sheets but any sprite.

**Reply by The-DevBlog** [reply]
I'm a newb at Bevy myself so forgive me if I'm not making sense haha

Can't you retrieve that info from a query? Eg:

```rust
fn my_function(
    mut commands: Commands,
    mut player_query: Query<(&Sprite, With<Player>)>,
) {
    if let Ok((player_sprite, _player)) = player_query.single_mut() 
    {
        let player_size_x = player_sprite.size.x;
        let player_size_y = player.sprite.size.y:
    }
}

```

Maybe I'm not fully comprehending what you mean by "after the sprite is spawned".

**Reply by The-DevBlog** [reply]
And if you would like to get the info from an individual frame, you could create a struct that holds the amount of columns and rows for a spritesheet, and then insert that to any entity you need to be able to get the single frame size. 

```rust 
struct SpriteSheetGrid 
{
    rows: u32,
    columns: u32
}
```

From here it would be some simple math to get a single frame.

**Reply by HotSammiches** [reply]
Yeah, I figured I could get it from a query after it was spawned but I was wondering if there was a way to do it in the same function where the sprite was spawned. I'm super new to Bevy and new to Rust so I'm still trying to learn the right mindset to make the best use of them both.

**Reply by The-DevBlog** [reply]
In that case, maybe create a struct to store the sprite sizes in and then insert that component/struct into your newly created entity and then you can query for that entity's component anywhere.

eg:

```rust
struct SpriteSize 
{
    height: f32,
    width: f32,
}

struct Player;

fn spawn_player(mut commands: Commands)
{
    let sprite_bundle = SpriteBundle {
        sprite: Sprite::new(Vec2::new(10.0, 10.0)),
        ..Default::default()
    };

    let sprite_size = SpriteSize {
        height: sprite_bundle.sprite.size.x,
        width: sprite_bundle.sprite.size.y  
    };

    commands.spawn_bundle(sprite_bundle).insert(Player).insert(sprite_size);
}
    
```

This way would make it easier to just query for the new struct `SpriteSize` like so:

```rust 
fn my_function(
    mut query: Query<(&SpriteSize, With<Player>)>
) 
{
    if let Ok((sprite_size, _player)) = query.single_mut() {
        let sprite_width = sprite_size.width;
        let sprite_height = sprite_size.height;

        // do something
    }
}
```

Either way you are still querying. I'm not aware of another way you would retrieve that kind of info.

**Reply by hymm** [reply]
The problem in the first snippet @HotSammiches posted is that assets are loaded asynchronously.
So when you run `let texture_handle = asset_server.load(PLAYER_SPRITE);`.  This is just telling the asset server to start loading the asset.  I think if you waited for the asset to be loaded, before running your spawning code you'd be able to get the sprite size.

For reference see https://bevy-cheatbook.github.io/cookbook/assets-ready.html

**Reply by HotSammiches** [reply]
I found another wrinkle here as I'm working on this problem. I call the startup_systems which load the assets and spawn the bundles and then call another system in StartupStage::PostStartup to update the global vector of collision boxes. I can't seem to figure out how to get the components which contain both a sprite and a CollisionBox component (one of my custom component types). Here's an example with the player sprite. First I call the routine which spawns the player bundle and inserts the collision box.

```
fn setup_player(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
    mut texture_atlases: ResMut<Assets<TextureAtlas>>,
) {
    let texture_handle = asset_server.load(PLAYER_SPRITE);
    let texture_atlas = TextureAtlas::from_grid(
        texture_handle,
        Vec2::new(PLAYER_SPRITE_TILE_SIZE_X, PLAYER_SPRITE_TILE_SIZE_Y),
        PLAYER_SPRITE_TILE_COLS,
        PLAYER_SPRITE_TILE_ROWS,
    );
    let texture_atlas_handle = texture_atlases.add(texture_atlas);

    commands
        .spawn_bundle(SpriteSheetBundle {
            texture_atlas: texture_atlas_handle,
            transform: Transform {
                scale: Vec3::splat(6.),
                rotation: Quat::from_rotation_x(0.),
                translation: Vec3::new(0., 0., 1.),
            },
            ..Default::default()
        })
        .insert(CollisionBox {
            ..Default::default()
        });
}
```

Then in the PostStartup function I run a query to find components with both a CollisionBox and a Transform. That function looks like this:

```
fn find_collision_boxen_with_transform(query: Query<(&CollisionBox, &Transform)>) {
    for (cb, sp) in query.iter() {
        println!("Found it!");
    }
}
```

This yields what I want - I get two lines printed since there are two entities with collision boxes I spawn right now. This is all fine. The problem is if I run a query on Sprite instead of Transform like this:

```
fn find_collision_boxen_with_sprite(query: Query<(&CollisionBox, &Sprite)>) {
    for (cb, sp) in query.iter() {
        println!("Found it!");
    }
}
```

This doesn't print anything. What am I missing?
